### DISCUSSION TITLE
Soliciting input regarding the hardening of a potential data leak

### DISCUSSION THREAD

**Discussion by nhorman** [openssl/openssl#24321] [General Discussion]
**The background**
Some research was recently sent to the openssl project regarding the potential leaking of data on the stack.

On some architectures that make use of various SIMD instructions (specifically AMD Zen4, but other arches may be affected), it is possible that, when various copy functions are used (like memcpy, strcpy, etc).  Its possible that, even if the source memory is cleared after the copy, data from the copy may remain in some of the simd registers, spilling onto the stack if lazy linking is performed.  This spillage could be benign, but it could also be potentially detrimental if the data in the memcpy operation is sensitive

**The analysis**
the problem, in an academic sense, is somewhat easy to fix.  Some compilers contain an attribute which allows the clearing of used simd (ne all) registers used in a function on exit.  It would be somewhat straightforward to apply this attribute to functions performing such copies and solve the problem that way.  But that approach has several drawbacks:
1) Its opportunistic - It works on platforms that support the attribute, but not all compilers do, so its effectively best effort
2) Its incomplete - It cleans up users of those functions, but leaves open the possibility of other locations that don't opt into that copy function, both within openssl and applications linking against it.  It also doesn't account for locations that might have the compiler emit instructions using simd registers as part of any optimization
3) Its a performance impact - While we've not measured the performance impact yet, it seems clear that, when used in a general sense, constantly clearing registers during copy operations is going to impact performance

**Additional information**
it should be noted that this problem was found outside of openssl.  There is no specific known attack on openssl using this discovery.  It also should be noted that exploitation of this would be a side channel attack, requiring local access to the system.  As such this is being looked at as a hardening effort, rather than a security bug.

**The question**
We are soliciting input for what to do about this.  The options currently are:

1) Augment our OPENSSL_memcpy/dup/strcpy functions to clear registers on exit, and opportunistically mitigate the problem where possible, accepting the performance impact

2) Create variants of the above functions with register clearing attributes set, and make use of them in places where we think sensitive data might be in a register, updating when/if additional locations are found that may leak sensitive data

**Comment by taladar** [comment]
I don't really see 1 and 2 as reasons not to do it in situations that are under your control. The fact that it does not cover all possible compilers or code paths using code not written by you is no reason to avoid fixing it in situations where it can be fixed that way.

As for 3, is there really a big performance impact beyond the actual clearing instruction if the register contains a value that is likely not used once the copy operation concludes? I don't really see this incurring additional memory loads if all you have to do is throw away the current register content that is not going to be used later anyway.

**Comment by kroeckx** [comment]
I think it would be good that we have a list of options that are available. You talk about an option to clear simd registers with some compilers, but I can't find that.

gcc has recently added support for clearing the stack: https://gcc.gnu.org/onlinedocs/gcc-14.1.0/gcc/Common-Type-Attributes.html#index-strub-type-attribute

**Reply by MartinBonner** [reply]
First utterly pedantic comment:  "its" is the possessive of "it", just as "his" is the possessive of "he".  "It's" on the other hand, means "It is", and is what is required above.

There is also a mixup in the first and second sentence of the second para.
> On some architectures that make use of various SIMD instructions (specifically AMD Zen4, but other arches may be affected), it is possible that, when various copy functions are used (like memcpy, strcpy, etc) ~~. Its possible~~ that, even if the source memory is cleared after the copy, data from the copy may remain in some of the simd registers, spilling onto the stack if lazy linking is performed.

However that results in a sentence which is much too long.  I suggest rewording as:

> Some architectures make use of various SIMD instructions (specifically AMD Zen4, but other arches may be affected) in various copy functions  (like memcpy, strcpy, etc).  On such architectures, it is possible that even if the source memory is cleared after the copy, data from the copy may remain in some of the simd registers, spilling onto the stack if lazy linking is performed.

**Reply by nhorman** [reply]
@kroeckx sorry, was referring to the zero_call_used_regs function attribute here, which should ideally clear the simd registers as well as any gprs.

**Comment by MartinBonner** [comment]
Actual substantive comment:  Surely the _first_ thing to do is to try enabling the relevant options and actually _measure_ the performance difference?  Until we know whether the cost is 50%, 10%, 1%, or 0%, it is impossible to have a sensible discussion about the appropriate approach.  (Note: 0% is possible if the CPU can clear the registers while it waits for some slower operation - like fetching the next instruction from the return address.)

**Reply by nhorman** [reply]
I'm personally less concerned about the specific performance impact, and more concerned about the trade off of any performance impact in exchange for a security benefit that is suspect at best.  the problem as described is a side channel attack, which is normally considered outside of scope for a security issue, and the current priority for openssl, based on the review of our issue backlog is performance.

So the question might better be asked as "is this sort of attack pragmatically a real concern for any users, given that it might involve some performance regression"

**Reply by MartinBonner** [reply]
The first priority for OpenSSL must always be "security".  I don't accept that a side-channel attack is not an attack.  I am also not convinced this issue cannot be accessed remotely - a further timing side-channel attack might disclose the spilled secrets (although this is clearly getting pretty complex).

So my answer to your rephrased question is "It might be, and we can't have a sensible discussion about whether to do anything about it until we know what the performance impact might be".
